### ISSUE TITLE
Faceted search

### ISSUE THREAD

**Issue by brady77** [dimonomid/geekmarks#13]
First of all: geekmarks are simply a great tool! Tags hierarchy is an ideal solution to union the world of 
- folders (to have something in one folder only, even if structured)
- flat tags (+ an external brain to remember all the tags in vocabulary)

I wish there was a faceted search:
- automatically generated **tag list** of _related tags_ just below the search output (_related tags_ are those tags that are applied to bookmarks and are part of the search output)
- clicking on a tag in the _tag list_ will add this tag to the search filter as well (lets call if "focused context")
- boolean logic for filtering: +boss+urgent-project1’, to find entries with tags ‘boss’ and ‘urgent’, but not ‘project1’, or ‘Kathy|Sally’ to find entries tagged as ‘Kathy’ or ‘Sally’ (as in [org-mode](https://orgmode.org/manual/Tag-searches.html#Tag-searches))

**Comment by dimonomid** [comment]
I'm kinda of two minds about the _related tags_ feature: e.g. I often apply some geo-location tags, like the name of the city, along with basically any other tag. E.g. if I bookmark a page about ergonomic chairs which are present in showroom in the city X, I'll tag the page with at least `furniture/chair`, `shopping`, and `city_X`. That tag `city_X` doesn't really seem "related" to `furniture/chair` in any way outside of that particular bookmark, so I'm not sure I want to see it as a suggestion when I create new tags.

Boolean logic is something I keep thinking myself occasionally. Sooner or later, one hopes... :)

**Comment by brady77** [comment]
I see your point. Let's say I have several bookmarks from my "furniture research". Some of them are tagged `[/furniture/chair, /shopping, /place/city_X]`, other bookmarks will be tagged `[/furniture/table, /shopping, /place/city_Y]`. If I want to review my research now I would probably search for `[/furniture, /shopping]`. For this I will use the _search dialog_ with three sections: **top**, **left** and **right**: 

On the **left** there will be a table with bookmarks matching my search criteria (Page Title, URL, Logo or Preview, Comment, applied tags).

On the **right** there will be an overview of relevant tags that are connected to the listed bookmarks (from the left section). According to our example search above you will see there `[/furniture/chair, /furniture/table, /shopping, /place/city_X, /place/city_Y]`.

The **top section** will have a _search input_ field and _tags filter_ field. The _search input_ will operate exactly as it works now: if I type "furn" I get hints `[/furniture, /furniture/chair, /furniture/table]`. If I click on a hint it will be added to the _tags filter_. As per our example the _tags filter_ will have `[/furniture, /shopping]`, now. 

Now I want to narrow my context to chairs only. So I go to the **right** side and I click on `[/furniture/chair]` tag. This will be added to the _tags filter_ on the **top**. Listed bookmarks on the **left** will be updated, as well as the listed tags on the **right**. If I want to broaden my context now to any location, I just go to the **top** and remove the tag `[/place/]` from the _tags filter_.

This scenario will probably prevent us from listing bookmarks that are "out of research context". For sure, if I remove all tags from _tags filter_ all but `[/place/city_X]`, my output will be populated with bookmarks about furniture and gardening. But this is something everyone should understand. Nothing prevents you from adding a tag `[/research/furniture-for-my-house`] to all the above mentioned bookmarks. That way you never loose the context at all.

Sorry for my complicated description.

**Comment by dimonomid** [comment]
Hmm, ok. Sorry I actually misunderstood you in the beginning: I was thinking that you're suggesting adding related tags when _adding_ a new bookmark, like, "would you like to also tag with X and Y?"

What you're suggesting does make sense.

**Comment by brady77** [comment]
Yes, exactly. This is actually "just" a better search and filter capability for your system. Sorry, I wasn't much transparent first. The boolean logic could be somehow added to the _ tags filter_ section. But don't know whether it will collide with the workflow described in my previous post.
